eth/tracers/native/erc7562.go

```diff
@@ -0,0 +1,550 @@
+}
+
+func (t *erc7562Tracer) OnTxEnd(receipt *types.Receipt, err error) {
+	defer catchPanic()
+	// Error happened during tx validation.
+	if err != nil {
+		return
+	}
+	t.callstackWithOpcodes[0].GasUsed = receipt.GasUsed
+	if t.config.WithLog {
+		// Logs are not emitted when the call fails
+		t.clearFailedLogs(&t.callstackWithOpcodes[0], false)
+	}
+}
+
+func (t *erc7562Tracer) OnLog(log1 *types.Log) {
+	defer catchPanic()
+	// Only logs need to be captured via opcode processing
+	if !t.config.WithLog {
+		return
+	}
+	// Skip if tracing was interrupted
+	if t.interrupt.Load() {
+		return
+	}
+	l := callLog{
+		Address:  log1.Address,
+		Topics:   log1.Topics,
+		Data:     log1.Data,
+		Position: hexutil.Uint(len(t.callstackWithOpcodes[len(t.callstackWithOpcodes)-1].Calls)),
+	}
+	t.callstackWithOpcodes[len(t.callstackWithOpcodes)-1].Logs = append(t.callstackWithOpcodes[len(t.callstackWithOpcodes)-1].Logs, l)
+}
+
+// GetResult returns the json-encoded nested list of call traces, and any
+// error arising from the encoding or forceful termination (via `Stop`).
+func (t *erc7562Tracer) GetResult() (json.RawMessage, error) {
+	defer catchPanic()
+	if len(t.callstackWithOpcodes) != 1 {
+		return nil, errors.New("incorrect number of top-level calls")
+	}
+
+	callFrameJSON, err := json.Marshal(t.callstackWithOpcodes[0])
+	if err != nil {
+		return nil, err
+	}
+
+	// Unmarshal the generated JSON into a map
+	var resultMap map[string]interface{}
+	if err := json.Unmarshal(callFrameJSON, &resultMap); err != nil {
+		return nil, err
+	}
+
+	// Converting keccak mapping to array
+	keccakArray := make([]hexutil.Bytes, len(t.Keccak))
+	i := 0
+	for k := range t.Keccak {
+		keccakArray[i] = hexutil.Bytes(k)
+		i++
+	}
+	resultMap["keccak"] = keccakArray
+
+	// Marshal the final map back to JSON
+	finalJSON, err := json.Marshal(resultMap)
+	if err != nil {
+		return nil, err
+	}
+	return finalJSON, t.reason
+}
+
+// Stop terminates execution of the tracer at the first opportune moment.
+func (t *erc7562Tracer) Stop(err error) {
+	defer catchPanic()
+	t.reason = err
+	t.interrupt.Store(true)
+}
+
+// clearFailedLogs clears the logs of a callframe and all its children
+// in case of execution failure.
+func (t *erc7562Tracer) clearFailedLogs(cf *callFrameWithOpcodes, parentFailed bool) {
+	failed := cf.failed() || parentFailed
+	// Clear own logs
+	if failed {
+		cf.Logs = nil
+	}
+	for i := range cf.Calls {
+		t.clearFailedLogs(&cf.Calls[i], failed)
+	}
+}
+
+func (t *erc7562Tracer) OnOpcode(pc uint64, op byte, gas, cost uint64, scope tracing.OpContext, rData []byte, depth int, err error) {
+	defer catchPanic()
+	opcode := vm.OpCode(op)
+	var opcodeWithStack *opcodeWithPartialStack
+	stackSize := len(scope.StackData())
+	var stackTopItems []uint256.Int
+	for i := 0; i < t.config.StackTopItemsSize && i < stackSize; i++ {
+		stackTopItems = append(stackTopItems, *peepStack(scope.StackData(), i))
+	}
+	opcodeWithStack = &opcodeWithPartialStack{
```

Comment:
I would switch on the opcode here instead of calling into multiple methods that might or might not do anything